crypto/evp_extra/evp_extra_test.cc

```diff
@@ -1123,6 +1123,79 @@ TEST(EVPExtraTest, Ed25519Keygen) {
                                reinterpret_cast<const uint8_t *>("hello"), 5));
 }
 
+// Test that |EVP_DigestSignFinal| and |EVP_DigestSignVerify| works with a
+// a special use case of not using the one-shot |EVP_DigestSignInit| or
+// |EVP_DigestVerifyInit| to initialize the |EVP_PKEY_CTX| context. The context
+// data can be manually constructed using other context setting functions.
+// |EVP_MD_CTX_set_pkey_ctx| was added to support this use case.
+TEST(EVPExtraTest, PKEY_CTX_manual) {
+  bssl::ScopedEVP_MD_CTX ctx;
+  ASSERT_TRUE(EVP_DigestInit(ctx.get(), EVP_sha256()));
+  ASSERT_TRUE(EVP_DigestUpdate(ctx.get(), kMsg, sizeof(kMsg)));
+
+  bssl::UniquePtr<EVP_PKEY> pkey = LoadExampleRSAKey();
+  ASSERT_TRUE(pkey);
+  bssl::UniquePtr<EVP_PKEY_CTX> pctx(EVP_PKEY_CTX_new(pkey.get(), nullptr));
+  ASSERT_TRUE(pctx);
+
+  // Manual construction for signing.
+  ASSERT_TRUE(EVP_PKEY_sign_init(pctx.get()));
+  ASSERT_TRUE(EVP_PKEY_CTX_set_signature_md(pctx.get(), EVP_sha256()));
+  EVP_MD_CTX_set_pkey_ctx(ctx.get(), pctx.get());
```

Comment:
In this call, should we verify that `ctx` and `pctx` has the same associated hash function?

In theory, you could associate different hash functions to each object. Might be nice if we can prevent that from happening. It would lead to an obvious error when trying to finalise the sign (at least for RSA, the EC implementation doesn't care).